File: cpp/KalkCpp/Shape/Shape2D/Polygon/Triangle/righttriangle.cpp
#include "righttriangle.h"
#include "math.h"
#include "typeinfo"


RightTriangle::RightTriangle(const std::vector<Point2D>& vp2d, ColorRGB* color): Triangle(vp2d, "Right Triangle", color)
{
    if(!canBuild(vp2d))
    {
        points = buildRightTriangle(1,1);
    }
}

RightTriangle* RightTriangle::clone() const
{
    return new RightTriangle(*this);
}

bool RightTriangle::canBuild(double cMax, double cMin) const
{
    return cMax > 0 && cMin > 0;
}

bool RightTriangle::canBuild(const std::vector<Point2D>& vp2d) const
{
    if(vp2d.size() == 3)
    {
        double base = Point2D::distance(vp2d[1], vp2d[0]);
        double height = Point2D::distance(vp2d[2], vp2d[0]);
        if(base > 0 && height > 0)
        {
            return true;
        }
    }
    return false;
}

std::vector<Point2D> RightTriangle::buildRightTriangle(double cathetusMax, double cathetusMin) const
{

    if(!canBuild(cathetusMax, cathetusMax))
    {
        cathetusMax = 1;
        cathetusMin = 1;
    }

    std::vector<Point2D> points = std::vector<Point2D>();

    points.push_back(Point2D());
    points.push_back(Point2D(cathetusMin));
    points.push_back(Point2D(0, cathetusMax));

    return points;
}



RightTriangle::RightTriangle(double cathetusMax, double cathetusMin, ColorRGB* color): Triangle(buildRightTriangle(cathetusMax, cathetusMin), "Right Triangle", color)
{

}


RightTriangle* RightTriangle::operator-(const Shape2D& s) const
{
    double newCathetusMin = 0;
    double newCathetusMax = 0;
    RightTriangle* newRightTriangle = nullptr;
    if(typeid(s) == typeid(const RightTriangle&))
    {
        const RightTriangle& rt = static_cast<const RightTriangle&>(s);
        newCathetusMin = base() - rt.base();
        newCathetusMax = height() - rt.height();
        newRightTriangle = new RightTriangle(buildRightTriangle(newCathetusMax, newCathetusMin), getColor()->operator -(*s.getColor()));
    }else{
        newCathetusMin = base() - s.sideToOperator();
        newCathetusMax = height() - s.sideToOperator();
        newRightTriangle = new RightTriangle(buildRightTriangle(newCathetusMax, newCathetusMin), getColor()->operator -(*s.getColor()));
    }
    return newRightTriangle;
}

RightTriangle* RightTriangle::operator+(const Shape2D& s) const
{
    double newCathetusMin = 0;
    double newCathetusMax = 0;
    RightTriangle* newRightTriangle = nullptr;
    if(typeid(s) == typeid(const RightTriangle&))
    {
        const RightTriangle& rt = static_cast<const RightTriangle&>(s);
        newCathetusMin = base() + rt.base();
        newCathetusMax = height() + rt.height();
        newRightTriangle = new RightTriangle(buildRightTriangle(newCathetusMax, newCathetusMin), getColor()->operator +(*s.getColor()));
    }else{
        newCathetusMin = base() + s.sideToOperator();
        newCathetusMax = height() + s.sideToOperator();
        newRightTriangle = new RightTriangle(buildRightTriangle(newCathetusMax, newCathetusMin), getColor()->operator +(*s.getColor()));
    }
    return newRightTriangle;
}

RightTriangle* RightTriangle::operator*(const Shape2D& s) const
{
    double newCathetusMin = 0;
    double newCathetusMax = 0;
    RightTriangle* newRightTriangle = nullptr;
    if(typeid(s) == typeid(const RightTriangle&))
    {
        const RightTriangle& rt = static_cast<const RightTriangle&>(s);
        newCathetusMin = base() * rt.base();
        newCathetusMax = height() * rt.height();
        newRightTriangle = new RightTriangle(buildRightTriangle(newCathetusMax, newCathetusMin), getColor()->operator *(*s.getColor()));
    }else{
        newCathetusMin = base() * s.sideToOperator();
        newCathetusMax = height() * s.sideToOperator();
        newRightTriangle = new RightTriangle(buildRightTriangle(newCathetusMax, newCathetusMin), getColor()->operator *(*s.getColor()));
    }
    return newRightTriangle;
}

RightTriangle* RightTriangle::operator/(const Shape2D& s) const
{
    double newCathetusMin = 0;
    double newCathetusMax = 0;
    RightTriangle* newRightTriangle = nullptr;
    if(typeid(s) == typeid(const RightTriangle&))
    {
        const RightTriangle& rt = static_cast<const RightTriangle&>(s);
        newCathetusMin = base() / rt.base();
        newCathetusMax = height() / rt.height();
        newRightTriangle = new RightTriangle(buildRightTriangle(newCathetusMax, newCathetusMin), getColor()->operator /(*s.getColor()));
    }else{
        newCathetusMin = base() / s.sideToOperator();
        newCathetusMax = height() / s.sideToOperator();
        newRightTriangle = new RightTriangle(buildRightTriangle(newCathetusMax, newCathetusMin), getColor()->operator /(*s.getColor()));
    }
    return newRightTriangle;
}


double RightTriangle::sideToOperator() const
{
    return base() + height();
}



double RightTriangle::base() const
{
    return Point2D::distance(getPoints()[1], getPoints()[0]);
}



double RightTriangle::height() const
{
    return Point2D::distance(getPoints()[2], getPoints()[0]);
}



double RightTriangle::hypotenuse() const
{
    return sqrt(pow(base(), 2) + pow(height(), 2));
}

double RightTriangle::area() const
{
    return (base() * height())/2;
}

File: Java/src/Shape/Shape2D/Polygon/Triangle/RightTriangle.java
package Shape.Shape2D.Polygon.Triangle;

import Color.ColorRGB;
import Shape.Point2D;
import Shape.Shape2D.Shape2D;

import java.util.ArrayList;

public class RightTriangle extends Triangle {

    public RightTriangle(ArrayList<Point2D> p) {
        super(p, "RightTriangle");
        if (!canBuild(p)) {
            points = buildRightTriangle(1, 1);
        }
    }

    public RightTriangle(ArrayList<Point2D> p, ColorRGB color) {
        super(p, "RightTriangle", color);
        if (!canBuild(p)) {
            points = buildRightTriangle(1, 1);
        }
    }

    public RightTriangle(double cathetusMin, double cathetusMax) {
        super(null, "RightTriangle");
        points = buildRightTriangle(cathetusMin, cathetusMax);
    }

    public RightTriangle(double cathetusMin, double cathetusMax, ColorRGB color) {
        super(null, "RightTriangle", color);
        points = buildRightTriangle(cathetusMin, cathetusMax);
    }

    private boolean canBuild(double cathetusMin, double cathetusMax) {
        return cathetusMin > 0 && cathetusMax > 0;
    }

    private boolean canBuild(ArrayList<Point2D> p) {
        if (p.size() == 3) {
            double cathetusMin = Point2D.distance(p.get(1), p.get(0));
            double cathetusMax = Point2D.distance(p.get(2), p.get(0));
            return cathetusMin > 0 && cathetusMax > 0;
        }
        return false;
    }

    private ArrayList<Point2D> buildRightTriangle(double cathetusMin, double cathetusMax) {
        if (!canBuild(cathetusMin, cathetusMax)) {
            cathetusMin = 1;
            cathetusMax = 1;
        }

        ArrayList<Point2D> points = new ArrayList<>();
        points.add(new Point2D());
        points.add(new Point2D(cathetusMin, 0));
        points.add(new Point2D(cathetusMin / 2, cathetusMax));

        return points;
    }

    public double sideToOperator() {
        return base() + height();
    }

    public double base() {
        return Point2D.distance(points.get(1), points.get(0));
    }

    public double height() {
        return Point2D.distance(points.get(2), new Point2D(base() / 2, 0));
    }

    public double hypotenuse() {
        return Math.sqrt(Math.pow(base(), 2) + Math.pow(height(), 2));
    }

    public double area() {
        return (base() * height()) / 2;
    }

    public RightTriangle sum(Shape2D other) {
        double newCathetusMin;
        double newCathetusMax;
        RightTriangle newRightTriangle;
        if (getClass() == other.getClass()) {
            RightTriangle it = (RightTriangle) other;
            newCathetusMin = base() + it.base();
            newCathetusMax = height() + it.height();
            newRightTriangle = new RightTriangle(buildRightTriangle(newCathetusMin, newCathetusMax)
                    , getColor().sum(other.getColor()));
        } else {
            newCathetusMin = base() + other.sideToOperator();
            newCathetusMax = height() + other.sideToOperator();
            newRightTriangle = new RightTriangle(buildRightTriangle(newCathetusMin, newCathetusMax)
                    , getColor().sum(other.getColor()));
        }
        return newRightTriangle;
    }

    public RightTriangle sub(Shape2D other) {
        double newCathetusMin;
        double newCathetusMax;
        RightTriangle newRightTriangle;
        if (getClass() == other.getClass()) {
            RightTriangle it = (RightTriangle) other;
            newCathetusMin = base() - it.base();
            newCathetusMax = height() - it.height();
            newRightTriangle = new RightTriangle(buildRightTriangle(newCathetusMin, newCathetusMax)
                    , getColor().sub(other.getColor()));
        } else {
            newCathetusMin = base() - other.sideToOperator();
            newCathetusMax = height() - other.sideToOperator();
            newRightTriangle = new RightTriangle(buildRightTriangle(newCathetusMin, newCathetusMax)
                    , getColor().sub(other.getColor()));
        }
        return newRightTriangle;
    }

    public RightTriangle mul(Shape2D other) {
        double newCathetusMin;
        double newCathetusMax;
        RightTriangle newRightTriangle;
        if (getClass() == other.getClass()) {
            RightTriangle it = (RightTriangle) other;
            newCathetusMin = base() * it.base();
            newCathetusMax = height() * it.height();
            newRightTriangle = new RightTriangle(buildRightTriangle(newCathetusMin, newCathetusMax)
                    , getColor().mul(other.getColor()));
        } else {
            newCathetusMin = base() * other.sideToOperator();
            newCathetusMax = height() * other.sideToOperator();
            newRightTriangle = new RightTriangle(buildRightTriangle(newCathetusMin, newCathetusMax)
                    , getColor().mul(other.getColor()));
        }
        return newRightTriangle;
    }

    public RightTriangle div(Shape2D other) {
        double newCathetusMin;
        double newCathetusMax;
        RightTriangle newRightTriangle;
        if (getClass() == other.getClass()) {
            RightTriangle it = (RightTriangle) other;
            newCathetusMin = base() / it.base();
            newCathetusMax = height() / it.height();
            newRightTriangle = new RightTriangle(buildRightTriangle(newCathetusMin, newCathetusMax)
                    , getColor().div(other.getColor()));
        } else {
            newCathetusMin = base() / other.sideToOperator();
            newCathetusMax = height() / other.sideToOperator();
            newRightTriangle = new RightTriangle(buildRightTriangle(newCathetusMin, newCathetusMax)
                    , getColor().div(other.getColor()));
        }
        return newRightTriangle;
    }

    public String toString() {
        return super.toString() + "CathetusMin: " + base() + "\nCathetusMax: " + height() + "\nHypotenuse: " +
                hypotenuse() + "\nArea: " + area() + "\nPerimeter: " + perimeter();
    }
}

